### PR TITLE
doc(pubsub): deprecate `PublisherOptions`

### PR DESCRIPTION
**Pull request by coryan**
Recommend using `google::cloud::Options` instead. As usual, I documented
the class as deprecated, but did not add a `[[deprecated]]` attribute
for the class, just for each member function.

Part of the work for #7348

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8911)

### CODE REVIEW

File: google/cloud/pubsub/publisher.h [hunk 1]
```diff
@@ -189,6 +185,19 @@ class Publisher {
     connection_->ResumePublish({std::move(ordering_key)});
   }
 
+  /// @deprecated Use `Publisher(connection)` instead.
+  GOOGLE_CLOUD_CPP_DEPRECATED("use `Publisher(connection)` instead")
+  explicit Publisher(std::shared_ptr<PublisherConnection> connection,
+                     std::initializer_list<internal::NonConstructible>)
+      : Publisher(std::move(connection)) {}
```
Comment (dbolduc) [inline google/cloud/pubsub/publisher.h:192]: optional: At the moment, this is unnecessary, as `Publisher(conn, {})` will hit the `PublisherOptions` overload. It only comes into play once we add `Publisher(std::shared_ptr<PublisherConnection> connection, Options options)`.

We can keep it though.

File: google/cloud/pubsub/publisher_options.cc [hunk 2]
```diff
@@ -20,6 +20,10 @@ namespace cloud {
 namespace pubsub {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 
+PublisherOptions::PublisherOptions() {
```
Comment (dbolduc) [inline google/cloud/pubsub/publisher_options.cc:23]: nit: we could save these lines by only offering a `Publisher(Options opts = {})`. This is fine though.

File: google/cloud/pubsub/publisher_options.h [hunk 3]
```diff
@@ -86,7 +93,13 @@ class PublisherOptions {
    * to [quota limits].
    *
    * [quota limits]: https://cloud.google.com/pubsub/quotas#resource_limits
+   *
```
Comment (dbolduc) [inline google/cloud/pubsub/publisher_options.h:96]: nit: delete this line?

**Review by dbolduc — APPROVED** [review]

File: google/cloud/pubsub/publisher.h (hunk 1 above)
Comment (coryan) [reply google/cloud/pubsub/publisher.h]: Removed.

File: google/cloud/pubsub/publisher_options.h (hunk 3 above)
Comment (coryan) [reply google/cloud/pubsub/publisher_options.h]: Done.

File: google/cloud/pubsub/publisher_options.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/pubsub/publisher_options.cc:23]: Will keep as-is, mostly laziness.

**Pull request merged**